File: PFAnalisisAlgoritmosPython/algoritmos/WinogradOriginal.py
# N -> columnas de la matriz B
# P -> columnas de la matriz A y filas de la matriz B
# M -> columnas de la matriz B

def WinogradOriginal(A, B, Result, N, P, M):
    upsilon = P % 2
    gamma = P - upsilon

    y = [0.0] * M
    z = [0.0] * N

    for i in range(M):
        aux = 0.0
        for j in range(0, gamma, 2):
            aux += A[i][j] * A[i][j + 1]
        y[i] = aux

    for i in range(N):
        aux = 0.0
        for j in range(0, gamma, 2):
            aux += B[j][i] * B[j + 1][i]
        z[i] = aux

    if upsilon == 1:
        PP = P - 1
        for i in range(M):
            for k in range(N):
                aux = 0.0
                for j in range(0, gamma, 2):
                    aux += (A[i][j] + B[j + 1][k]) * (A[i][j + 1] + B[j][k])
                Result[i][k] = aux - y[i] - z[k] + A[i][PP] * B[PP][k]
    else:
        for i in range(M):
            for k in range(N):
                aux = 0.0
                for j in range(0, gamma, 2):
                    aux += (A[i][j] + B[j + 1][k]) * (A[i][j + 1] + B[j][k])
                Result[i][k] = aux - y[i] - z[k]

    return Result

File: PFAnalisisAlgoritmosGo/algoritmos/WinogradOriginal.go
package algoritmos

// WinogradOriginal realiza la multiplicación de matrices usando el algoritmo de Winograd.
func WinogradOriginal(A, B, Result [][]float64, N, P, M int) [][]float64 {
	upsilon := P % 2
	gamma := P - upsilon

	y := make([]float64, M)
	z := make([]float64, N)

	// Calcular el vector y
	for i := 0; i < M; i++ {
		aux := 0.0
		for j := 0; j < gamma; j += 2 {
			aux += A[i][j] * A[i][j+1]
		}
		y[i] = aux
	}

	// Calcular el vector z
	for i := 0; i < N; i++ {
		aux := 0.0
		for j := 0; j < gamma; j += 2 {
			aux += B[j][i] * B[j+1][i]
		}
		z[i] = aux
	}

	// Realizar la multiplicación dependiendo del valor de upsilon
	if upsilon == 1 {
		PP := P - 1
		for i := 0; i < M; i++ {
			for k := 0; k < N; k++ {
				aux := 0.0
				for j := 0; j < gamma; j += 2 {
					aux += (A[i][j] + B[j+1][k]) * (A[i][j+1] + B[j][k])
				}
				Result[i][k] = aux - y[i] - z[k] + A[i][PP]*B[PP][k]
			}
		}
	} else {
		for i := 0; i < M; i++ {
			for k := 0; k < N; k++ {
				aux := 0.0
				for j := 0; j < gamma; j += 2 {
					aux += (A[i][j] + B[j+1][k]) * (A[i][j+1] + B[j][k])
				}
				Result[i][k] = aux - y[i] - z[k]
			}
		}
	}

	return Result
}
